zcash_client_sqlite/src/wallet/init/migrations/transparent_gap_limit_handling.rs

```diff
@@ -0,0 +1,492 @@
+        // duplicative of information that can be discovered via joins with the various
+        // `*_received_{notes|outputs}` tables, which we will create a view to perform below.
+        conn.execute_batch(&format!(
+            r#"
+            CREATE TABLE addresses_new (
+                id INTEGER NOT NULL PRIMARY KEY,
+                account_id INTEGER NOT NULL,
+                key_scope INTEGER NOT NULL DEFAULT {external_scope_code},
+                diversifier_index_be BLOB NOT NULL,
+                address TEXT NOT NULL,
+                transparent_child_index INTEGER,
+                cached_transparent_receiver_address TEXT,
+                exposed_at_height INTEGER,
+                FOREIGN KEY (account_id) REFERENCES accounts(id),
+                CONSTRAINT diversification UNIQUE (account_id, key_scope, diversifier_index_be),
+                CONSTRAINT transparent_index_consistency CHECK (
+                    (transparent_child_index IS NOT NULL) == (cached_transparent_receiver_address IS NOT NULL)
+                )
+            );
+
+            INSERT INTO addresses_new (
+                account_id, key_scope, diversifier_index_be, address,
+                transparent_child_index, cached_transparent_receiver_address
+            )
+            SELECT
+                account_id, key_scope, diversifier_index_be, address,
+                transparent_child_index, cached_transparent_receiver_address
+            FROM addresses;
+            "#
+        ))?;
+
+        // Now, we add the ephemeral addresses to the newly unified `addresses` table.
+        #[cfg(feature = "transparent-inputs")]
+        {
+            let mut ea_insert = conn.prepare(
+                r#"
+                INSERT INTO addresses_new (
+                    account_id, key_scope, diversifier_index_be, address,
+                    transparent_child_index, cached_transparent_receiver_address
+                ) VALUES (
+                    :account_id, :key_scope, :diversifier_index_be, :address,
+                    :transparent_child_index, :cached_transparent_receiver_address
+                )
+                "#,
+            )?;
+
+            let mut ea_query = conn.prepare(
+                r#"
+                SELECT account_id, address_index, address
+                FROM ephemeral_addresses
+                "#,
+            )?;
+            let mut rows = ea_query.query([])?;
+            while let Some(row) = rows.next()? {
+                let account_id: i64 = row.get("account_id")?;
+                let transparent_child_index = row.get::<_, i64>("address_index")?;
+                let diversifier_index = DiversifierIndex::from(
+                    u32::try_from(transparent_child_index).map_err(|_| {
+                        WalletMigrationError::CorruptedData(
+                            "ephermeral address indices must be in the range of `u32`".to_owned(),
```

Comment:
Maybe go via `NonHardenedChildIndex` before `DiversifierIndex` to detect any "`u32` but not `u31`" corruption.